### ISSUE TITLE
Tailwindcss2 @apply directive doesn't work inside vue component

### ISSUE THREAD

**Issue by ShekhSaifuddin007** [JeffreyWay/laravel-mix#2778]
-   Laravel Mix Version:  "^6.0.6"
-   Node Version: 14.4.0
-   NPM Version: 6.14.5
-   OS: windows

I create a laravel application with jetstream and inertia-vue stack for my new project problem is Tailwindcs version 2 using `postCss` and it doesn't support `@apply` directive inside vue components but inside .css file it works fine I don't want that because that css file will load my every page I just want short inline utility classes with `@apply` directive but I can't, How Can I achieve that.?

### inside my vue template

```html
<template>
 <div class="mt-4">
  <label for="hello">Hello</label>
  <input id="hello" class="input"/>
 </div>
</templete>

<style scoped>
    .input {
        @apply bg-gray-200 border h-10
    }
</style>
```

#### output inside browser like this

![Capture](https://user-images.githubusercontent.com/48278861/104767441-581df000-5796-11eb-9f23-2df7adcef354.PNG)

#### webpack.mix.js

```js
const mix = require('laravel-mix');

mix.js('resources/js/app.js', 'public/js').vue()
    .postCss('resources/css/app.css', 'public/css', [
        require('postcss-import'),
        require('tailwindcss'),
        require('autoprefixer'),
    ])
    .webpackConfig(require('./webpack.config'));

if (mix.inProduction()) {
    mix.version();
}
```

#### webpack.config.js

```js
const path = require('path');

module.exports = {
    resolve: {
        alias: {
            '@': path.resolve('resources/js'),
        },
    },
};
```

**Comment by nivv** [comment]
Have you tried below?
```html
<style lang="postcss" scoped>
    .input {
        @apply bg-gray-200 border h-10
    }
</style>
```
I can't get  `@apply` to work in my scoped styles either.

**Comment by nivv** [comment]
I got it working after installing `laravel-mix-tailwind` and using it this way in my `webpack.config.js`

```js
require('laravel-mix-tailwind')

mix.postCss('resources/css/app.css', 'public/css', [
    require('tailwindcss')('./tailwind.config.js')
]).tailwind(
```

Style example in my vue file (notice the absence of the lang attribute:

```html
<style>
.lookie {
    @apply bg-pink-100 text-pink-600 rounded-full border border-pink-400 mb-8;
}
</style>
```

**Comment by thecrypticace** [comment]
Yeah this is something we need to look at. At the moment you have to specify the postCSS plugins globally for it to affect anything other than the specific css/sass/less/etc… file/tree you've given via `mix.postCss` / et al:

```js
mix.options({
    postCss: [require("tailwindcss")("./tailwind.config.js")]
})
```

Note that you can set these plugins globally in your mix file like above and do not need to specify plugins via mix.postCss. You also have the option of using a postcss config file which we will honor as well.

@JeffreyWay What are your thoughts on this? Should we leave this as is or do you think it'd be reasonable to populate the PostCSS plugins globally from a single `mix.postCss` call. I'm on the fence about it.

**Comment by fylzero** [comment]
> I got it working after installing `laravel-mix-tailwind` and using it this way in my `webpack.config.js`
> 
> ```js
> require('laravel-mix-tailwind')
> 
> mix.postCss('resources/css/app.css', 'public/css', [
>     require('tailwindcss')('./tailwind.config.js')
> ]).tailwind()
> ```
> 
> Style example in my vue file (notice the absence of the lang attribute:
> 
> ```
> <style>
> .lookie {
>     @apply bg-pink-100 text-pink-600 rounded-full border border-pink-400 mb-8;
> }
> </style>
> ```

Couldn't get this to install.  Looks like maybe it doesn't support Laravel Mix 6 yet?

**Comment by nivv** [comment]
> > I got it working after installing `laravel-mix-tailwind` and using it this way in my `webpack.config.js`
> > ```js
> > require('laravel-mix-tailwind')
> > 
> > mix.postCss('resources/css/app.css', 'public/css', [
> >     require('tailwindcss')('./tailwind.config.js')
> > ]).tailwind()
> > ```
> > 
> > 
> > Style example in my vue file (notice the absence of the lang attribute:
> > ```
> > <style>
> > .lookie {
> >     @apply bg-pink-100 text-pink-600 rounded-full border border-pink-400 mb-8;
> > }
> > </style>
> > ```
> 
> Couldn't get this to install. Looks like laravel-mix-tailwind doesn't support Laravel Mix 6 yet?
> 
> ```
> npm ERR! code ERESOLVE
> npm ERR! ERESOLVE unable to resolve dependency tree
> npm ERR!
> npm ERR! Found: laravel-mix@6.0.10
> npm ERR! node_modules/laravel-mix
> npm ERR!   dev laravel-mix@"^6.0.6" from the root project
> npm ERR!
> npm ERR! Could not resolve dependency:
> npm ERR! dev laravel-mix-tailwind@"*" from the root project
> npm ERR!
> npm ERR! Conflicting peer dependency: laravel-mix@2.1.14
> npm ERR! node_modules/laravel-mix
> npm ERR!   peer laravel-mix@"^2.1.7" from laravel-mix-tailwind@0.1.1
> npm ERR!   node_modules/laravel-mix-tailwind
> npm ERR!
> npm ERR! Fix the upstream dependency conflict, or retry
> npm ERR! this command with --force, or --legacy-peer-deps
> npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
> npm ERR!
> npm ERR! See /Users/myuseraccount/.npm/eresolve-report.txt for a full report.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     /Users/myuseraccount/.npm/_logs/2021-01-18T15_50_22_169Z-debug.log
> ```

Hm, that's strange. I'm using yarn, maybe that's why. I installed `laravel-mix-tailwind` before `laravel-mix`

**Comment by nivv** [comment]
So now I have the use case of using postcss in a vue component like below:

```vue
<style lang="postcss">
.test {
    @apply bg-red-400; /*Not working */
}
```
I've also added this to my webpack config:

```
        rules: [
            {
                test: /\.postcss$/,
                use: [
                    'vue-style-loader',
                    'css-loader',
                    {
                        loader: 'postcss-loader'
                    }
                ]
            }
        ]
```

So basically I have to choose between tailwind or postcss, which is weird. I though `@apply` was parsed with postcss anyways?

**Comment by ShekhSaifuddin007** [comment]
If wanna use postcss and you wanna use `@apply` directive to work inside vue components then just install this plugin and inside your `webpack.mix.js` configure like this, don't need to do anything


### for postcss

```js
const mix = require('laravel-mix');
require('laravel-mix-tailwind')

mix.js('resources/js/app.js', 'public/js').vue()
    .postCss('resources/css/app.css', 'public/css', [
        require('postcss-import'),
        require('tailwindcss')('./tailwind.config.js'),
        require('autoprefixer'),
    ])
    .webpackConfig(require('./webpack.config'))
    .tailwind();

if (mix.inProduction()) {
    mix.version();
}
```

if you wanna use a preprocessor like sass etc.. then you don't need to install this plugin, and then uninstall `postcss`,  `postcss-import`, `autoprefixer` from `package.json` and install `resolve-url-loader`, `sass`, `sass-loader` and your `webpack.mix.js` will look like this

`npm uninstall postcss postcss-import autoprefixer` then `npm install --save-dev resolve-url-loader sass sass-loader`

### for preprocessor
```js
const mix = require('laravel-mix');
const tailwindcss = require('tailwindcss');

mix.js('resources/js/app.js', 'public/js').vue()
    .sass('resources/sass/app.scss', 'public/css')
    .options({
        processCssUrls: false,
        postCss: [ tailwindcss('./tailwind.config.js') ],
    })
    .webpackConfig(require('./webpack.config'));

if (mix.inProduction()) {
    mix.version();
}
```

**Hope guys this will help**

**Comment by fylzero** [comment]
@ShekhSaifuddin007 Are you able to install laravel-mix-tailwind on a Laravel 8 project?  I'm getting a conflicting peer dependency when I try (posted above).

**Comment by anotherglitchinthematrix** [comment]
The solution on the fresh install is as simple as putting a bare config file on the root of the project. [Source](https://tailwindcss.com/docs/installation#add-tailwind-as-a-post-css-plugin).

```js
// postcss.config.js
module.exports = {
    plugins: {
        tailwindcss: {},
        autoprefixer: {},
    },
}
```

**Comment by fylzero** [comment]
@anotherglitchinthematrix Created `./postcss.config.js` with the contents below...  didn't make a difference.  Same error.

```
module.exports = {
    plugins: {
        tailwindcss: {},
        autoprefixer: {},
    }
}
```

**Comment by fylzero** [comment]
> @ShekhSaifuddin007 Are you able to install laravel-mix-tailwind on a Laravel 8 project? I'm getting a conflicting peer dependency when I try (posted above).
> 
> This seems to be the culprit.
> 
> https://github.com/JeffreyWay/laravel-mix-tailwind/blob/master/package.json on line 25
> 
> ```
> "peerDependencies": {
>     "laravel-mix": "^2.1.7"
> }
> ```
> 
> I submitted this as an issue to that repo: [JeffreyWay/laravel-mix-tailwind#30](https://github.com/JeffreyWay/laravel-mix-tailwind/issues/30)

This has been resolved with the latest patch version of laravel-mix-tailwind.  

@see
https://github.com/JeffreyWay/laravel-mix-tailwind/issues/30#issuecomment-767678728

**Comment by ShekhSaifuddin007** [comment]
@fylzero I think probably need to update your `laravel-mix`

**Comment by KonradDeskiewicz** [comment]
So to sum up my experience: 
- can't use `lang="postcss"` in style (webpack throws errors then) and therefore get linter errors when using long `@apply` with many classes
- browserSync don't reflect changes in cue component style (yes, it recompiles "something" but no changes in the page)

I really  expected that `.vue().tailwind()` will solve the issue...

**Comment by deslittle** [comment]
I believe this may have been fixed in v6.0.16 which has upgraded the dependency "vue-style-loader": "^4.1.3".

**Comment by bebetoalves** [comment]
> I believe this may have been fixed in v6.0.16 which has upgraded the dependency "vue-style-loader": "^4.1.3".

It's not fixed.

**Comment by bebetoalves** [comment]
A workaround that can temporarily solve this problem without using laravel-mix-tailwind:

In `webpack.mix.js`:
```js
const mix = require('laravel-mix')

/*
 |--------------------------------------------------------------------------
 | Mix Asset Management
 |--------------------------------------------------------------------------
 |
 | Mix provides a clean, fluent API for defining some Webpack build steps
 | for your Laravel applications. By default, we are compiling the CSS
 | file for the application as well as bundling up all the JS files.
 |
 */

mix.ts('resources/js/app.ts', 'public/js').vue()
  .postCss('resources/css/app.css', 'public/css')
  .webpackConfig(require('./webpack.config'))
  .sourceMaps()

if (mix.inProduction()) {
  mix.version()
}

```

In `webpack.config.js`:
```js
const path = require('path')

module.exports = {
  resolve: {
    alias: {
      '@': path.resolve('resources/js')
    }
  },
  module: {
    rules: [
      {
        test: /\.(postcss)$/,
        use: [
          'vue-style-loader',
          { loader: 'css-loader', options: { importLoaders: 1 } },
          'postcss-loader'
        ]
      }
    ]
  }
}

``` 

In `postcss.config.js`:
```js
module.exports = {
  plugins: [
    require('postcss-import'),
    require('tailwindcss')('./tailwind.config.js'),
    require('autoprefixer')
  ]
}
```

Now I can use:
```vue
<style lang="postcss">
.input-form {
  @apply block w-full mt-1 border-gray-300 rounded-md shadow-sm focus:border-indigo-300 focus:ring focus:ring-indigo-200 focus:ring-opacity-50;
}
</style>
```

**Comment by olemarius** [comment]
I have the same issue. Added storybook 6.2.9 to a Vite+ vue 3 typescript project - everything works fine but I get an error when using the @apply directive. Also upgraded postcss to postcss 8. 

I've tried this approach https://blog.spell.codes/en/storybook-tailwind-scss/ but seems this only applyes to *.scss files.

So far I've been unable to find a way to add the loader appropriately. If someone find a good way to do this it should definitively be added to the docs.

**Comment by vortechron** [comment]
after a thousand google searches, finally, I got the answer from here: 

tq @nivv 

```
require('laravel-mix-tailwind');

mix.js('resources/js/app.js', 'public/js').vue()
    .postCss('resources/css/app.css', 'public/css', [
        require('postcss-import'),
        require('tailwindcss'),
    ])
    .webpackConfig(require('./webpack.config'))
    .tailwind();
```

**Comment by thecrypticace** [comment]
You can do this in a few ways:
1. Recommended way: Use a `postcss.config.js` file to configure your list of plugins. Mix will use these as well and you won't have to specify them in your webpack.mix.js file.
2. Alternatively you can call `mix.options({ postCss: [… plugin list here …] })`

It would be somewhat nice for a single call to mix.postCss to set the global plugin list but it would be unexpected if separate sets of plugins are used for different files and one of them got set for "global" usage.

I'll give this some thought.

**Comment by plakhin** [comment]
Can anyone tell finally what needs to be done to be able to use `@apply` directive inside a `.vue` file on the fresh Laravel 8 install?
What plugins need to be installed? What needs to be added into config(s)? Should one use `<style>` or `<style lang="postcss">`? What else?

**Comment by mshamaseen** [comment]
after applying @bebetoalves answer, I needed to remove the `land="postcss"` inside the vue style in order for it to work.

**Comment by ivodolenc** [comment]
Hi, I also bumped on this problem via google 😂 After a few hours I finally set up the working configuration.

This setup solves issues with Tailwind `@apply` directives inside Vue `<style></style>` section.

Just tested it and you can also use `postcss` via `<style lang="postcss"></style>` method.

### Setup Environment

- LaravelMix 6+
- Vue 3+
- Postcss 8+
- TailwindCSS 3+

### mix.config.js

```js
const { resolve } = require('path')
const mix = require('laravel-mix')

/**
 * Webpack Configuration
 */
mix.webpackConfig({
	module: {
		rules: [
			// Postcss settings for classic <style></style>
			{
				test: /\.css$/i,
				loader: 'postcss-loader',
				options: {
					postcssOptions: {
						config: resolve(__dirname, 'postcss.config.js') // Path to your postcss.config.js
					}
				}
			},
			// Postcss settings for <style lang="postcss"></style>
			{
				test: /\.postcss$/i,
				use: [
					'vue-style-loader',
					{
						loader: 'css-loader',
						options: { importLoaders: 1 }
					},
					{
						loader: 'postcss-loader',
						options: {
							postcssOptions: {
								config: resolve(__dirname, 'postcss.config.js') // Path to your postcss.config.js
							}
						}
					}
				]
			}
		]
	}
})
```

### postcss.config.js

[More info](https://github.com/webpack-contrib/postcss-loader#config)

```js
module.exports = {
	plugins: [
		require('postcss-import'),
		require('tailwindcss')('./tailwind.config.js'), // Path to your tailwind.config.js
		require('autoprefixer')
	]
}
```

### package.json

Required stack

```json
{
	"dependencies": {
		"vue": "^3.2.33"
	},
	"devDependencies": {
		"autoprefixer": "^10.4.5",
		"laravel-mix": "^6.0.43",
		"postcss": "^8.4.12",
		"postcss-import": "^14.1.0",
		"tailwindcss": "^3.0.24",
		"vue-loader": "^17.0.0"
	}
}
```

### Working examples

Everything works fine (including **BrowserSync**), you can even use `@apply` without `lang="postcss"` but then you will get a css warning `Unknown at rule @apply`.

```html
<style scoped>
	.text {
		color: blue;
	}
</style>
```

```html
<style lang="postcss" scoped>
	.text {
		@apply text-green-400;
	}
</style>
```

**Comment by thecrypticace** [comment]
@ivodolenc you do not need to set up all of this configuration to get it to work.

With Mix you have two fairly simple options
1. `mix.options({ postCss: [ require('tailwindcss')('./tailwind.config.js') ] })` — This applies the given PostCSS plugins to ALL css compiled in Mix.
2. Alternatively, if you have a `postcss.config.js` file you can put it there and Laravel Mix will pick up that file and use it for all compiled CSS as well.

**Comment by ivodolenc** [comment]
> With Mix you have two fairly simple options
>
> 1. mix.options({ postCss: [ require('tailwindcss')('./tailwind.config.js') ] }) — This applies the given PostCSS plugins to ALL css compiled in Mix.
> 2. Alternatively, if you have a postcss.config.js file you can put it there and Laravel Mix will pick up that file and use it for all compiled CSS as well.

Hmmm I think this will only import `TailwindCSS` into the project and it will work fine for utility classes.

But when you need to write custom css inside `Vue SFC` then it will throws some loader errors (due to missing postcss addition config).

```html
<!-- won't work without extra config -->

<style lang="postcss" scoped>
	.text {
		@apply text-green-400;
	}
</style>
```

Additional [config](https://github.com/laravel-mix/laravel-mix/issues/2778#issuecomment-1113962606) above is for those who want to use Tailwind `@apply` directives and `lang="postcss"` inside Vue `<style></style>` section.

Maybe I'm wrong but I only managed to run it if I specified an additional webpack configuration for `postcss-loader`.

**Comment by plakhin** [comment]
True, @apply directive doesn't work for me without additional config.

**Comment by thecrypticace** [comment]
You don't need `lang="postcss"` All Vue style blocks are processed by postcss already.

**Comment by ivodolenc** [comment]
@thecrypticace Ok thanks for feedback 👍

**Comment by repeater09** [comment]
I still cannot get this to work. I believe I have tried every suggestion on a fresh jetstream install. 

Can anyone provide a step-by-step that actually works?

**Comment by stephan-v** [comment]
I don't really understand why `lang="postcss"` is no longer needed because vue-loader no longer seems to apply postcss transforms automatically according to the documentation:

https://vue-loader.vuejs.org/guide/pre-processors.html#postcss

If Laravel Mix is doing something to enable this then I would also expect plugins that are being passed to `mix.postCss()` to also apply to the transformation that is being done on Vue component styles but this does not seem to be the case.

The only way I can properly get Tailwind working for single file Vue component is to add a global `postcss.config.js` file and at that point I can also remove the plugins that were previously being passed to the `mix.postCss()` method.

It seems broken at this moment because postcss transformations and plugins are applied but only to `.css` files. Plugins don't seem to work for Vue components.
